File: Moderate/Sorting and searching/MergeIntervals.java


/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }

 Example 1:

Input: [[1,3],[2,6],[8,10],[15,18]]
Output: [[1,6],[8,10],[15,18]]
Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
Example 2:

Input: [[1,4],[4,5]]
Output: [[1,5]]
Explanation: Intervals [1,4] and [4,5] are considerred overlapping.

 */
class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        
        List<Interval> result = new ArrayList<Interval>();
        if(intervals.size()==0)
        {
        	return result;
        }
        quickSort(intervals,0,intervals.size()-1);
        Interval currentInterval = new Interval(intervals.get(0).start,intervals.get(0).end);
        result.add(currentInterval);
        int resultIndex = 0;
        for(int i=1;i<intervals.size();i++)
        {
        	if(intervals.get(i).start>result.get(resultIndex).end)
        	{
        		resultIndex++;
        		Interval intrvl = new Interval(intervals.get(i).start,intervals.get(i).end);
        		result.add(resultIndex,intrvl);
        	}
        	else if(intervals.get(i).end>result.get(resultIndex).end)
        	{
        		result.get(resultIndex).end = intervals.get(i).end;
        	}
        }
        return result;
    }

    private void quickSort(List<Interval> intervals, int start, int end)
    {
    	if(start<end)
    	{
    		int q = partition(intervals,start,end);
    		quickSort(intervals,start,q-1);
    		quickSort(intervals,q+1,end);
    	}
    }
    private int partition(List<Interval> intervals,int start, int end)
    {
    	int key = intervals.get(end).start;
    	int leftIndex = start;
    	for(int i=start;i<end;i++)
    	{
    		if(intervals.get(i).start<key)
    		{
    			Collections.swap(intervals,i,leftIndex);
    			leftIndex++;
    		}
    	}
    	Collections.swap(intervals,end,leftIndex);
		return leftIndex;
    }
}

File: CSharp/Medium/Sorting and Searching/MergeIntervals.cs
/*
Example 1:

Input: [[1,3],[2,6],[8,10],[15,18]]
Output: [[1,6],[8,10],[15,18]]
Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
Example 2:

Input: [[1,4],[4,5]]
Output: [[1,5]]
Explanation: Intervals [1,4] and [4,5] are considerred overlapping.
*/
/**
 * Definition for an interval.
 * public class Interval {
 *     public int start;
 *     public int end;
 *     public Interval() { start = 0; end = 0; }
 *     public Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
	private int HeapSize =0;
    public IList<Interval> Merge(IList<Interval> intervals) {
    	IList<Interval> result = new List<Interval>();
    	
    	if(intervals.Count==0)
    	{
    		return result;
    	}
    	Sort(intervals);
    	int currentStart = intervals[0].start;
        int currentEnd = intervals[0].end;
        
        for(int i=1;i<intervals.Count;i++)
        {
        	if(intervals[i].start<=currentEnd)
        	{
        		if(intervals[i].end>currentEnd)
        		{
        			currentEnd = intervals[i].end;
        		}
        	}
        	else
        	{
        		Interval intrvl = new Interval(currentStart,currentEnd);
        		result.Add(intrvl);
        		currentStart = intervals[i].start;
        		currentEnd = intervals[i].end;
        	}
        }
        Interval interval = new Interval(currentStart,currentEnd);
        result.Add(interval);
        return result;
    }

    private void Sort(IList<Interval> intervals)
    {
    	BuildHeap(intervals);
    	for(int i=0;i<intervals.Count;i++)
    	{
    		Swap(intervals,0,HeapSize-1);
    		HeapSize--;
    		MaxHeapify(intervals,0);
    	}
    }

    private void BuildHeap(IList<Interval> intervals)
    {
    	HeapSize = intervals.Count;
    	for(int i=intervals.Count/2;i>=0;i--)
    	{
    		MaxHeapify(intervals,i);
    	}
    }

    private void MaxHeapify(IList<Interval> intervals, int index)
    {
    	int left = index*2+1;
    	int right = index*2+2;
    	int max = index;
    	if(left<HeapSize && intervals[left].start>intervals[index].start)
    	{
    		max = left;
    	}
    	if(left<HeapSize && intervals[left].start==intervals[index].start && intervals[left].end>intervals[index].end)
    	{
    		max = left;
    	}
    	if(right<HeapSize && intervals[right].start>intervals[max].start)
    	{
    		max = right;
    	}
    	if(right<HeapSize && intervals[right].start==intervals[max].start && intervals[right].end>intervals[max].end)
    	{
    		max = right;
    	}
    	if(max!=index)
    	{
    		Swap(intervals,index,max);
    		MaxHeapify(intervals,max);
    	}
    }

    private void Swap(IList<Interval> intervals, int i, int j)
    {
    	Interval temp = intervals[i];
    	intervals[i] = intervals[j];
    	intervals[j] = temp;
    }
}